### PR TITLE
refactor(CL): move id management and pool creation to swaprouter

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #3097

## What is the purpose of the change

This PR moves pool id management to `swaprouter` from `gamm`. This will allow sharing the logic for managing ids between `gamm` and `concentrated-liquidity` modules.

To enable this, we will have a unified entry point for creating pools - `CreatePoolMsg`. It has also been moved from `gamm` to `swaprouter` together with tests.

## Changelog

- move the `next_pool_id` index from `gamm` to `swaprouter`
- create a new index in gamm for `pool_count`
- port tests and queries to `swaprouter`
- refactor tests and queries in `gamm` to be compatible with `pool_count` abstraction
- implement parameter migrations in the upgrade handler
- port and fix simulation tests 
- fix e2e tests that covers upgrade and migration logic

## TODOs
- [x] issue for: https://github.com/osmosis-labs/osmosis/blob/0c2fa2d14a153b50d5d97525556ad828d3af6344/x/swaprouter/creator_test.go#L37-L39
- [ ] browse TODOs in swaprouter and create issues
- [ ] browse comments and create issues

### CODE REVIEW

File: x/concentrated-liquidity/pool.go
Comment (mattverse) [inline x/concentrated-liquidity/pool.go:17]: let's remove this for now?

File: x/concentrated-liquidity/pool.go
Comment (mattverse) [inline x/concentrated-liquidity/pool.go:39]: I don't think this should be returning a poolI from the Gamm types right?

File: x/gamm/client/testutil/test_helpers.go
Comment (mattverse) [inline x/gamm/client/testutil/test_helpers.go:113]: Should we move exit pool to the swap router as well?

File: x/gamm/module.go
Comment (czarcas7ic) [inline x/gamm/module.go:68]: why this change?

File: x/gamm/keeper/keeper.go
Comment (mattverse) [inline x/gamm/keeper/keeper.go:83]: should we create an issue to come back and migrate the params?

File: x/gamm/keeper/export_test.go
Comment (mattverse) [inline x/gamm/keeper/export_test.go:19]: I think this isn't an efficient way of storing given that we already have the poolCount that we have to set to(since we're rewriting and storing every iteration). I think we should create a new method called`setPoolCount` that could set the value directly

File: x/gamm/keeper/pool.go
Comment (mattverse) [inline x/gamm/keeper/pool.go:53]: can we move the check to the very top so that we don't do unnecessary operations before failing

File: x/swaprouter/types/params.go
Comment (czarcas7ic) [inline x/swaprouter/types/params.go:28]: this is swaprouter module parameters right?

File: x/gamm/keeper/pool.go
Comment (mattverse) [inline x/gamm/keeper/pool.go:20]:
```suggestion
	// Add the share token's meta data to the bank keeper.
```

File: x/gamm/types/key.go
Comment (mattverse) [inline x/gamm/types/key.go:29]: why not unify the usage when they're same values?

File: x/concentrated-liquidity/pool.go
Comment (p0mvn) [reply x/concentrated-liquidity/pool.go:39]: I'm planning to move the interfaces to the `swaprouter` later but keep the implementations per-module

File: x/swaprouter/client/query_proto_wrap.go
Comment (mattverse) [inline x/swaprouter/client/query_proto_wrap.go:113]: Consider changing the name from `NumPools` to `NextPoolId` since we potentially have plans to remove pools and then this query name could be misleading

File: x/gamm/client/testutil/test_helpers.go
Comment (p0mvn) [reply x/gamm/client/testutil/test_helpers.go:111]: yeah this is ported from gamm. Since the pool creation fee now exists in `swaprouter`, we need to port this substitution

File: x/swaprouter/creator.go
Comment (mattverse) [inline x/swaprouter/creator.go:88]: Let's move MinPoolAssets and MaxPoolAssets to the swaprouter types since they're not being used or called anywhere else in the gamm module

File: x/swaprouter/creator_test.go
Comment (mattverse) [inline x/swaprouter/creator_test.go:57]: do you also want to add and test for the things that are happening in the `CreatePool` flow? ex. check that funds are actually moving to the community pool, check that lp has been minted to the account, check that module account has been created etc)

File: x/swaprouter/types/params.go
Comment (mattverse) [inline x/swaprouter/types/params.go:28]:
```suggestion
// default swaprouter module parameters.
```

File: x/swaprouter/types/msg_create_pool.go
Comment (mattverse) [inline x/swaprouter/types/msg_create_pool.go:22]: Would it be possible to have TraditionalAMMInterface live inside the swap router module?

File: x/swaprouter/types/listeners.go
Comment (mattverse) [inline x/swaprouter/types/listeners.go:10]: Can we name them hooks instead of listeners?

**Review by mattverse — COMMENTED** [review]
Left some reviews! 
Please take a look 🌮

File: x/swaprouter/types/listeners.go
Comment (p0mvn) [reply x/swaprouter/types/listeners.go:10]: According to discussions with @ValarDragon during `twap` implementation, we are going to go towards pub-sub listeners in the future. So it might make sense to start adjusting our abstractions to listeners incrementally

File: x/swaprouter/types/msg_create_pool.go
Comment (p0mvn) [reply x/swaprouter/types/msg_create_pool.go:22]: This is the goal of a future refactor

File: x/swaprouter/creator_test.go
Comment (p0mvn) [reply x/swaprouter/creator_test.go:57]: Yeah agreed I don't want to be blocking this PR on this though since this is a straight up port

File: app/keepers/keepers.go
Comment (p0mvn) [reply app/keepers/keepers.go:270]: The are all utilized by the pool creation logic

File: x/gamm/module.go
Comment (p0mvn) [reply x/gamm/module.go:68]: Since pool id is removed, there is no genesis to validate anymore

File: x/twap/listeners.go
Comment (p0mvn) [reply x/twap/listeners.go:14]: addressed

File: x/twap/listeners.go
Comment (p0mvn) [reply x/twap/listeners.go:46]: Same reply as here: https://github.com/osmosis-labs/osmosis/pull/3170#discussion_r1006217084

File: proto/osmosis/gamm/v1beta1/genesis.proto
Comment (p0mvn) [reply proto/osmosis/gamm/v1beta1/genesis.proto]: fixed

File: x/concentrated-liquidity/pool.go
Comment (p0mvn) [reply x/concentrated-liquidity/pool.go:17]: This must be defined because the interface for the concentrated-liquidity module field in swap router keeper requires it

File: x/concentrated-liquidity/pool.go
Comment (p0mvn) [reply x/concentrated-liquidity/pool.go:40]: We will migrate pool interface to a better location in a future PR

File: x/gamm/client/testutil/test_helpers.go
Comment (p0mvn) [reply x/gamm/client/testutil/test_helpers.go]: Moved

File: x/gamm/keeper/keeper.go
Comment (p0mvn) [reply x/gamm/keeper/keeper.go:83]: What do you mean by migrate? They are completely removed. I don't see any logic to explicitly clear param state in the `x/params` module

Comment (p0mvn) [reply x/gamm/keeper/keeper.go:83]: If you mean migrate to swaprouter, then the only parameter is set in the upgrade handler

File: x/gamm/keeper/export_test.go
Comment (p0mvn) [reply x/gamm/keeper/export_test.go]: This was removed, not even sure what this is

File: x/gamm/types/key.go
Comment (p0mvn) [reply x/gamm/types/key.go]: I created a separate index for pool count, going to deprecate and remove pool id completely in a future release

File: x/swaprouter/client/query_proto_wrap.go
Comment (p0mvn) [reply x/swaprouter/client/query_proto_wrap.go]: Seems that we are keeping pools. Also, I don't think we need to have a query for pool ids. I don't think any client would be interested in that. What clients might want to know is how many pools there are which we can infer from the next pool id

Comment (p0mvn) [reply x/swaprouter/client/query_proto_wrap.go]: Also, this is a direct port from gamm. I'm not being creative here. Let me know if you feel strongly though, we can discuss further

File: x/gamm/keeper/pool.go
Comment (czarcas7ic) [inline x/gamm/keeper/pool.go:174]: Seems like it sets it to a user defined number, not 0

File: x/swaprouter/keeper.go
Comment (czarcas7ic) [inline x/swaprouter/keeper.go:96]:
```suggestion
// SetPoolCreationListeners sets the pool creation listeners.
func (k *Keeper) SetPoolCreationListeners(listeners types.PoolCreationListeners) *Keeper {
```

**Review by czarcas7ic — COMMENTED** [review]
No glaring issues, looks good to merge once we get CI passing

File: app/app.go
Comment (p0mvn) [reply app/app.go:425]: This was done for debugging e2e upgrade failures only. Going to remove

File: proto/osmosis/gamm/v1beta1/genesis.proto
Comment (p0mvn) [reply proto/osmosis/gamm/v1beta1/genesis.proto:14]: We can't simply remove a field. We can only mark it as "reserved", indicating that it is not used anymore.

File: wasmbinding/stargate_whitelist.go
Comment (p0mvn) [reply wasmbinding/stargate_whitelist.go]: Under both. Updated the missing one

File: x/gamm/client/cli/cli_test.go
Comment (mattverse) [inline x/gamm/client/cli/cli_test.go:47]: Are these to move in the future as well?

**Review by mattverse — COMMENTED** [review]
Do you have a written spec of some kind some where? (Doesnt have to be formal, having hard time reviewing or getting a understanding of how the wiring would look like in the future steps + confusion on what would remain permenant and what would be moved in future PRs)

**Review by czarcas7ic — APPROVED** [review]

File: x/gamm/keeper/pool.go
Comment (p0mvn) [reply x/gamm/keeper/pool.go:166]: Sure. It might be useful to know the number of pools in `gamm`, in `concentrated-liquidity and in total in `swaprouter`. This is what updates the counter in gamm.

File: x/gamm/pool-models/stableswap/pool.go
Comment (p0mvn) [reply x/gamm/pool-models/stableswap/pool.go:411]: Agreed this is to be addressed in a follow-up issue

File: app/upgrades/v13/upgrades.go
Comment (p0mvn) [reply app/upgrades/v13/upgrades.go:33]: We are testing the upgrade to v13 in e2e test. To make sure it passes we need to have migration logic somewhere. We will just end up moving it to v14 upgrade handler when v13 is released

**Comment by p0mvn** [comment]
> Do you have a written spec of some kind some where? (Doesnt have to be formal, having hard time reviewing or getting a understanding of how the wiring would look like in the future steps + confusion on what would remain permenant and what would be moved in future PRs)

The core of the change is moving pool creation to the swaprouter:
https://github.com/osmosis-labs/osmosis/blob/b9984996088ed1d5ad38fbf512aa9d1dd48a17db/x/swaprouter/creator.go

The rest is simply moving tests and refactoring to lift the pool creation and id management to the swaprouter.

`swaprouter`'s `CreatePool` directs pool creation to `InitializePool` implemented by:
- gamm: https://github.com/osmosis-labs/osmosis/blob/b9984996088ed1d5ad38fbf512aa9d1dd48a17db/x/gamm/keeper/pool.go#L17
- concentrated-liquidity: https://github.com/osmosis-labs/osmosis/blob/b9984996088ed1d5ad38fbf512aa9d1dd48a17db/x/concentrated-liquidity/pool.go#L15

This is to be added to: https://github.com/osmosis-labs/osmosis/pull/3052

File: proto/osmosis/gamm/v1beta1/genesis.proto
Comment (p0mvn) [reply proto/osmosis/gamm/v1beta1/genesis.proto:14]: Copied over and missed the update, changed now. Should be 2

File: x/gamm/pool-models/stableswap/pool.go
Comment (p0mvn) [reply x/gamm/pool-models/stableswap/pool.go:411]: The reason why this hasn't been moved right now is because I'm limiting the scope to only doing the absolutely necessary lifts for the refactor.

Otherwise, this PR would get out of control, reaching a few thousand more LOC. Sorry if this was confusing

File: x/concentrated-liquidity/pool.go
Comment (p0mvn) [reply x/concentrated-liquidity/pool.go]: Removed it

**Comment by p0mvn** [comment]
@mattverse spec updates: https://github.com/osmosis-labs/osmosis/pull/3233

File: proto/osmosis/gamm/v1beta1/genesis.proto
Comment (czarcas7ic) [reply proto/osmosis/gamm/v1beta1/genesis.proto]: I thought you said we can remove fields in major upgrades, just not minor?

Comment (p0mvn) [reply proto/osmosis/gamm/v1beta1/genesis.proto]: We can remove them. However, we need to mark the number as reserved like here:
https://github.com/osmosis-labs/osmosis/blob/2733c688c88276a1ef641064a70920dfc92f869f/proto/osmosis/epochs/genesis.proto#L63

**Comment by p0mvn** [comment]
This is good to go, discussed follow-up items offline and created / creating issues

**Pull request merged**
